cli/core/test/e2e/framework/plugin_operations.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package framework
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	plugintypes "github.com/vmware-tanzu/tanzu-framework/apis/cli/v1alpha1"
+)
+
+// PluginOps helps to perform plugin operations like create plugin bundle, and publish
+type PluginOps interface {
+
+	// GenerateScriptBasedPluginBinaries generates script based plugin binaries for given plugin metadata and return generated plugin binary file paths
+	GenerateScriptBasedPluginBinaries(pluginsMeta []*PluginMeta) ([]string, []error)
+
+	// PublishPluginBinary publishes the plugin binaries to given registry bucket and returns the plugin distribution urls
+	PublishPluginBinary(pluginsInfo []*PluginMeta) (distributionUrls []string, errs []error)
+
+	// GeneratePluginBundle generates plugin bundle in local file system for given plugin metadata
+	GeneratePluginBundle(pluginsMeta []*PluginMeta) ([]string, []error)
+
+	// PublishPluginBundle publishes the plugin bundles to given registry bucket and returns the plugins discovery urls
+	PublishPluginBundle(pluginsInfo []*PluginMeta) (discoveryUrls []string, errs []error)
+
+	// AddPluginDiscoverySource adds plugin discovery source
+	AddPluginDiscoverySource(discoveryOpts *DiscoveryOptions) error
+
+	// ListPlugins lists all plugins by running 'tanzu plugin list' command
+	ListPlugins() ([]PluginListInfo, error)
```

Comment:
The way these two functions are implemented for `localOCIPluginOps` is using the `tanzu plugin list` and `tanzu plugin source add` commands, it doesn't look like this should be part of this interface. This interface is for publishing plugins to a specific endpoint (local OCI repository in this case) and shouldn't include plugin lifecycle operations.

`ListPlugins` is a generic API that is not associated with any specific type of pluginOps. There might be multiple plugin sources added to the config file and when I do `tanzu plugin list` it will return the plugins from all the plugin sources. So, I feel we need to have a different interface to execute plugin lifecycle commands as part of the framework.
And This interface should be limited to configuring a single plugin discovery source.